code/build/android-native/java/org/orx/lib/OrxNativeActivity.java

```diff
@@ -0,0 +1,139 @@
+package org.orx.lib;
+
+import android.annotation.TargetApi;
+import android.os.Build;
+import android.os.Bundle;
+import android.util.Log;
+import android.app.NativeActivity;
+import android.content.Context;
+import android.hardware.input.InputManager;
+import android.view.InputDevice;
+
+/**
+    Base class for the apps NativeActivity adapted to orxAndroidNativeSupport.cpp in orxlib.
+
+    NativeActivity is used with the android_native_app_glue.
+ */
+public class OrxNativeActivity extends NativeActivity implements InputManager.InputDeviceListener {
+
+    public static int orxANDROID_KU32_MAX_JOYSTICK_NUMBER = 16; // same as in Orx /code/include/io/orxJoystick.h
+
+    // C functions we call
+    public native void nativeOnInputDeviceAdded(int deviceId);
+    public native void nativeOnInputDeviceChanged(int deviceId);
+    public native void nativeOnInputDeviceRemoved(int deviceId);
+    private native void nativeOnPause();
+    private native void nativeOnResume();
+
+    private InputManager mInputManager;
+    private int mGameControllerIds[] = new int[orxANDROID_KU32_MAX_JOYSTICK_NUMBER];
+
+    /*
+      NOTE: Do NOT to call any native methods before App is created (before this callback is done).
+      The native app structure is initialized in new thread when super.onCreate() is being executed.
+      So best is to wait until onStart() below has been called.
+    */
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        mInputManager = (InputManager)this.getSystemService(Context.INPUT_SERVICE);
+    }
+
+    @Override
+    protected void onStart() {
+      Log.d("OrxNativeActivity", "onStart()");
+    	super.onStart();
+
+      getAndStoreDeviceIds();
+      mInputManager.registerInputDeviceListener(this, null);
+    }
+
+    @Override
+    protected void onStop() {
+      Log.d("OrxNativeActivity", "onStop()");
+      mInputManager.unregisterInputDeviceListener(this);

```

Comment:
Shouldn't this call be in `onPause()`?

Reply:
Adopted from the old code OrxActivity onStop() see above.

Reply:
> @znakeeye Did you get a chance to have a look?
> Is it OK for me to merge this PR in its current state?

I still don't understand why all this cannot be done in `C`, i.e. most changes should be in `android_native_app_glue.c`. Exactly which feature requires JNI/Java?

Reply:
If you read my long explanation from Feb 12 in the Conversation you can see why JNI is needed.
android_native_app_glue.c is provided by Android NDK so we have our extra glue in C as the orxAndroid(Native)Support-files.

Reply:
> If you read my long explanation from Feb 12 in the Conversation you can see why JNI is needed.
> android_native_app_glue.c is provided by Android NDK so we have our extra glue in C as the orxAndroid(Native)Support-files.

for me this good. @jonasohrn up to you to resolve this converstation